### PR TITLE
watchdb adaptive ac

### PR DESCRIPTION
**Pull request by lyh169**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/watcher/ac_cache.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		value := cc.l.Remove(elm)
+		delete(cc.m, version)
+		return value.(*MessageCacheEvent)
+	}
+	return nil
+}
+
+func (cc *commitCache) getTop() (*MessageCacheEvent, bool) {
+	cc.mtx.Lock()
+	defer cc.mtx.Unlock()
```
Comment (lcmmhcc) [inline x/evm/watcher/ac_cache.go]: should getTop use RLock?

File: x/evm/watcher/watcher.go [hunk 2]
```diff
@@ -362,6 +386,37 @@ func (w *Watcher) CommitWatchData(data WatchData) {
...
+func (w *Watcher) CommitWatchDataToCache(data WatchData) {
+	if data.Size() == 0 {
+		return
+	}
+	if data.Batches != nil {
+		w.acProcessor.BatchSetEx(data.Batches)
+	}
+	if data.DirtyList != nil {
+		w.acProcessor.BatchDel(data.DirtyList)
+	}
```
Comment (chengzhinei) [inline x/evm/watcher/watcher.go:454]: // todo put DelayEraseKey to BatchDel

File: x/evm/watcher/watcher.go [hunk 3]
```diff
@@ -480,7 +571,22 @@ func (w *Watcher) ApplyWatchData(watchData interface{}) {
...
+		shouldPersist := w.IsShouldPersist()
+		if shouldPersist {
+			w.acProcessor.MoveToCommitList(int64(w.height)) // move curMsgCache to commitlist
+		}
+		w.dispatchJob(func() {
+			w.AsyncCommitWatchData(wd, shouldPersist)
+		})
+	} else {
+		w.dispatchJob(func() {
+			w.CommitWatchData(wd)
```
Comment (chengzhinei) [inline x/evm/watcher/watcher.go:624]: w.ExecuteDelayEraseKey(data.DelayEraseKey)

**Comment by lyh169** [comment]
pr: https://github.com/okex/exchain/pull/2473

File: x/evm/watcher/ac_processor.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+func init() {
+	gACProcessor = &ACProcessor{
+		commitList:  newCommitCache(), // for support to querier
+		curMsgCache: newMessageCache(),
+	}
+}
+
+func SetAnableAsyncCommit(enable bool) {
```
Comment (giskook) [inline x/evm/watcher/ac_processor.go]:
```suggestion
func SetEnableAsyncCommit(enable bool) {
```

File: x/evm/watcher/ac_cache.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+import (
+	"container/list"
+	"encoding/hex"
+	"sync"
+)
+
+type MessageCache struct {
+	mtx sync.RWMutex
+	mp  map[string]WatchMessage // if the key of value WatchMessage is nil, this key should del on db batch write
```
Comment (giskook) [inline x/evm/watcher/ac_cache.go]: I saw a DelType below, Maybe the comment is misleading?

File: x/evm/watcher/watcher.go [hunk 6]
```diff
@@ -336,9 +374,52 @@ func (w *Watcher) Commit() {
...
+		w.dispatchJob(func() {
+			acBatch, noACBatch := ClassifyWatchMessageType(batch)
+			w.acProcessor.BatchSet(acBatch)
+			if shouldPersist {
+				// move curMsgCache to commitlist
+				w.acProcessor.MoveToCommitList(int64(height))
+				w.commitBatch(noACBatch)
+				w.AsyncCommitBatch()
+			} else {
+				w.commitBatch(noACBatch)
```
Comment (giskook) [inline x/evm/watcher/watcher.go]: Seems we can move `w.commitBath` out of the code block ,then remove the `else` code block.

File: x/evm/watcher/ac_processor.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+	ap.mtx.Lock()
+	ap.curMsgCache = newMessageCache()
+	ap.mtx.Unlock()
+}
+
+func (ap *ACProcessor) PersistHander(commitFn func(epochCache *MessageCache)) {
+	// commit to db
+	for {
+		cmmiter, ok := ap.commitList.getTop()
```
Comment (giskook) [inline x/evm/watcher/ac_processor.go]: Maybe rename  `cmmiter` to `committer`, `PersistHander` to `PersistHandler` will be better

File: x/evm/watcher/watcher.go [hunk 8]
```diff
@@ -623,13 +802,23 @@ func (w *Watcher) lazyInitialization() {
...
 		return
 	}
 	close(w.jobChan)
 	w.jobDone.Wait()
+	w.ACProcessorStop()
 }
+
+func (w *Watcher) ACProcessorStop() {
+	if GetEnableAsyncCommit() {
+		driveHeight := int64(w.height + 100) // use a greater height trigger last commit
```
Comment (giskook) [inline x/evm/watcher/watcher.go]: I saw `1<<63 - 1` is used by iavl/nodedb to identify the latest version.maybe we can use it here?

File: x/evm/watcher/ac_querier.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+	return nil, errACNotFound
+}
+
+func (aq *ACProcessorQuerier) GetAccount(key []byte) (*types.EthAccount, error) {
+	if v, ok := aq.p.Get(key); ok {
+		if v == nil || v.GetType() == TypeDelete { // for del key
+			return nil, nil
+		}
+		if rsp, ok := v.(*MsgAccount); ok {
+			cp := rsp.account.Copy() // the account value will be modify by caller
```
Comment (giskook) [inline x/evm/watcher/ac_querier.go]: the comment `be modify` to `be modified`

File: x/evm/watcher/ac_cache.go (hunk 5 above)
Comment (lyh169) [reply x/evm/watcher/ac_cache.go]: > I saw a DelType below, Maybe the comment is misleading?

yes

File: x/evm/watcher/ac_processor.go (hunk 4 above)
Comment (lyh169) [reply x/evm/watcher/ac_processor.go]: > 
ok

File: x/evm/watcher/ac_processor.go (hunk 7 above)
Comment (lyh169) [reply x/evm/watcher/ac_processor.go]: > Maybe rename `cmmiter` to `committer`, `PersistHander` to `PersistHandler` will be better

ok

File: x/evm/watcher/ac_querier.go (hunk 9 above)
Comment (lyh169) [reply x/evm/watcher/ac_querier.go]: > the comment `be modify` to `be modified`

ok

File: x/evm/watcher/watcher.go (hunk 6 above)
Comment (lyh169) [reply x/evm/watcher/watcher.go]: ok

File: x/evm/watcher/watcher.go (hunk 8 above)
Comment (lyh169) [reply x/evm/watcher/watcher.go]: > 1<<63 - 1

ok

**Review by giskook — DISMISSED** [review]

**Review by ilovers — DISMISSED** [review]

**Review by chengzhinei — DISMISSED** [review]

**Review by lcmmhcc — APPROVED** [review]

**Pull request closed without merging**